presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -56,6 +59,7 @@
         private final List<Type> groupByTypes;
         private final List<Integer> groupByChannels;
         private final List<Integer> globalAggregationGroupIds;
+        private final List<Integer> preGroupedChannels;
```

Comment:
this is clever but it makes me wonder if we will use this feature of `HashAggregationOperator`?

I was rather thinking of a dedicated streaming aggregation operator that would allow us to avoid hash lookups, hash computations and so on (e.g: similar as `AggregationOperator`). Such dedicated streaming aggregation operator should be more performant in correlated subqueries context than extended `HashAggregationOperator`

Reply:
@sopel39 @martint Karol, I'm seeing that we currently use `AggregationOperator` for global aggregations (set of grouping keys is empty) and `HashAggregationOperator` for all other aggregations. I'm also seeing that correlated subqueries generate aggregations with grouping keys made of `unique` and *all the probe columns*. Such aggregations prevent pruning of probe columns and require unnecessary hashing. Alternatively, sub-queries could be using aggregations on just the `unique` and apply `arbitrary` grouping function to the probe columns. E.g.

`select name, (select max(name) from region where regionkey = nation.regionkey) from nation;`

could be re-written to

```
- Aggregation [unique] -> arbitrary(nation.name), max(region.name)
   - LeftJoin (region.regionkey = nation.regionkey)
      - AssignUniqueId
         - TableScan(nation)
      - TableScan(region)
```

This way, the planner would see that aggregation input is grouped on *all* grouping columns and hashing is unnecessary and be able to plan streaming aggregation without hashing. Otherwise, if the planner only sees that aggregation input is grouped on a subset of grouping columns, it can't automatically decide whether simple aggregation or hash aggregation is better. What do you think?

Reply:
>  Alternatively, sub-queries could be using aggregations on just the unique and apply arbitrary grouping function to the probe columns. E.g.

I think such construct will confuse stats computation. We could improve `TransformCorrelatedScalarAggregationToJoin` though to not include unused symbols in top aggregation. I think for that we would need to extend `LateralJoinNode` to include explicit list of `outputSymbols` as `JoinNode` do.

Nota that pruning passes through `LateralJoinNode` but correlated (but unused later on) symbols won't be pruned from probe side and will take part in aggregation.

> This way, the planner would see that aggregation input is grouped on all grouping columns and hashing is unnecessary and be able to plan streaming aggregation without hashing

I'm confused. I think `AssignUniqueId` makes all columns grouped (not only `uniqueId` column). See: https://github.com/prestodb/presto/pull/10731/files/1b9e5c11be55febd887fc80bd1a6fbc68d08da26#diff-64cfd4c0cfe165b7dcc4934f126200e5